bug: Waku Store returns duplicate messages

### Problem

js-waku interop test:

1. Get nwaku to send 20 messages using `post_waku_v2_relay_v1_message` JSON RPC API
2. Retrieve messages using page size 10, backward direction

Expected: 20 messages are returned
Actual: 44 messages are returned


### Impact
Seems to be a critical bug until investigation is cleared.

### To reproduce

From a js-waku clone:
```
git checkout chore/nwaku-0.14
npm run i
npm run build
cd packages/tests
DEBUG=waku* TS_NODE_PROJECT=./tsconfig.dev.json mocha tests/store.node.spec.ts --grep "^Waku Store Generator$"
```

nwaku logs will be present in `./log` folder

Note nwaku is executed with the following arguments:

```
--listen-address=127.0.0.1 --rpc=true --relay=true --rpc-admin=true --websocket-support=true --log-level=TRACE --tcp-port=10840 --rpc-port=10839 --websocket-port=10841 --store=true --lightpush=true
```
and env var:

```
WAKUNODE2_STORE_MESSAGE_DB_URL = "";
```

### Screenshots/logs

nwaku logs: 
[nwaku_Waku_Store_Generator.log](https://github.com/waku-org/nwaku/files/10463560/nwaku_Waku_Store_Generator.log)


js-waku logs: 

```
  waku:nwaku nwaku args: --listen-address=127.0.0.1 --rpc=true --relay=true --rpc-admin=true --websocket-support=true --log-level=TRACE --tcp-port=34097 --rpc-port=34096 --websocket-port=34098 --store=true --lightpush=true +0ms
  waku:nwaku nwaku 958844 started at 4:50:15 pm +57ms
  waku:nwaku Waiting to see 'Node setup complete' in nwaku logs +0ms
  waku:nwaku nwaku node has been started +4s
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '00',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819306000000n
  }
] +4ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +44ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '01',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819351000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +9ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '02',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819361000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +9ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '03',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819370000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +8ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '04',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819378000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '05',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819386000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '06',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819394000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '07',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819401000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '08',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819409000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '09',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819417000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +6ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0a',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819424000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +6ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0b',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819431000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +6ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0c',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819438000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0d',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819445000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +8ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0e',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819453000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0f',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819461000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +6ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '10',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819467000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '11',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819475000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +6ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '12',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819482000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +7ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '13',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674193819489000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +6ms
  waku:waku Waku node created 12D3KooWAsQqEPpwncpMS9MMDkshKHzZWLPd4Xh1C94JxywkWUA5 relay: false, store: true, light push: true, filter: true, peer exchange: true  +0ms
  waku:nwaku RPC Query:  get_waku_v2_debug_v1_info [] +406ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":{"listenAddresses":["/ip4/127.0.0.1/tcp/34097/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi","/ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi"],"enrUri":"enr:-KO4QPJ54S-f-6rLtBWeE167Q-fmjqRpNQb6rmo0LQ8Vpj0NSVpC6bjHxvsVjKAlzAi8T9uiOQfZHwCVy6eUQFi74TUBgmlkgnY0gmlwhH8AAAGKbXVsdGlhZGRyc4wACgR_AAABBoUy3QOJc2VjcDI1NmsxoQKpZfGlPpAPEItcqIbIZoQGZVyWl4AnTV_y802TiyY0EYN0Y3CChTGFd2FrdTIL"}} +6ms
  waku:waku Dialing to /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi with protocols store,filter,lightpush,peer-exchange +147ms
  waku:wait-for-remote-peer /vac/waku/store/2.0.0-beta4 peer found:  16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi +0ms
  waku:store Querying history with the following options { peerId: undefined } +0ms
  waku:libp2p-utils Using codec /vac/waku/store/2.0.0-beta4 +0ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +53ms
  waku:store 5 messages retrieved from store +28ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +57ms
  waku:store 5 messages retrieved from store +56ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +60ms
  waku:store 5 messages retrieved from store +57ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +61ms
  waku:store 5 messages retrieved from store +55ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +58ms
  waku:store 5 messages retrieved from store +54ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +56ms
  waku:store 5 messages retrieved from store +57ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +59ms
  waku:store 5 messages retrieved from store +56ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +56ms
  waku:store 5 messages retrieved from store +54ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/34098/ws/p2p/16Uiu2HAm6pw3JzpCCJY8f6YPUgiur65R3Ze6vvnh5D5bjsEJFdmi for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +55ms
  waku:store 4 messages retrieved from store +60ms
  waku:store Stopping pagination due to `response.pagingInfo.cursor` missing from store response +0ms
[
  14, 15, 16, 17, 18, 12, 13, 14, 15, 16,
  10, 11, 12, 13, 14,  8,  9, 10, 11, 12,
   6,  7,  8,  9, 10,  4,  5,  6,  7,  8,
   2,  3,  4,  5,  6,  0,  1,  2,  3,  4,
   0,  1,  2,  3
]
  waku:nwaku nwaku 958844 getting SIGINT at 4:50:21 pm +1s
  waku:nwaku nwaku 958844 interrupted: true +0ms
  waku:nwaku nwaku 958844 process exited with 0 at 4:50:21 pm +24ms

AssertionError: expected 44 to equal 20
Expected :20
Actual   :44
<Click to see difference>

    at Context.<anonymous> (file:///home/fryorcraken/src/waku-org/js-waku/packages/tests/tests/store.node.spec.ts:93:30)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

```
[
  14, 15, 16, 17, 18, 12, 13, 14, 15, 16,
  10, 11, 12, 13, 14,  8,  9, 10, 11, 12,
   6,  7,  8,  9, 10,  4,  5,  6,  7,  8,
   2,  3,  4,  5,  6,  0,  1,  2,  3,  4,
   0,  1,  2,  3
]
```

Shows the dupe messages.


### nwaku version/commit hash

- e.g. `v0.14.0`

I expect the setup above to use the in-memory db. As nwaku uses sqlite by default, I have changed the js-waku test harness to also use sqlite. This resulted in the same behaviour.

commit that replaces the usage of in-memory db with sqlite: https://github.com/waku-org/js-waku/pull/1128/commits/60b7912a5c5bb2f81c6a1363352eed7361bc304a

nwaku logs: 
[nwaku_Waku_Store_Generator.log](https://github.com/waku-org/nwaku/files/10476475/nwaku_Waku_Store_Generator.log)


js-waku logs:
```
  waku:nwaku nwaku args: --listen-address=127.0.0.1 --rpc=true --relay=true --rpc-admin=true --websocket-support=true --log-level=TRACE --tcp-port=48773 --rpc-port=48772 --websocket-port=48774 --store=true --lightpush=true --store-message-db-url=sqlite://ygyqj.sqlite3 +0ms
  waku:nwaku nwaku 1593672 started at 2:11:05 pm +129ms
  waku:nwaku Waiting to see 'Node setup complete' in nwaku logs +1ms
  waku:nwaku nwaku node has been started +4s
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '00',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469497000000n
  }
] +5ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +63ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '01',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469561000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +15ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '02',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469576000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +11ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '03',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469588000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +11ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '04',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469599000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +11ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '05',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469610000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '06',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469620000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '07',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469630000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +9ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '08',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469640000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +11ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '09',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469652000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0a',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469662000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0b',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469673000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0c',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469683000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0d',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469694000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0e',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469704000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +8ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '0f',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469712000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +9ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '10',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469721000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '11',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469732000000n
  }
] +1ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +9ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '12',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469741000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:nwaku RPC Query:  post_waku_v2_relay_v1_message [
  '/waku/2/default-waku/proto',
  {
    payload: '13',
    contentTopic: '/test/1/waku-store/utf8',
    timestamp: 1674443469751000000n
  }
] +0ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":true} +10ms
  waku:waku Waku node created 12D3KooWA31yXnwcrMa4xdQZXs6GwgdDUZjUyoBrN4TLoSHmJdsZ relay: false, store: true, light push: true, filter: true, peer exchange: true  +0ms
  waku:nwaku RPC Query:  get_waku_v2_debug_v1_info [] +477ms
  waku:nwaku RPC Response:  {"jsonrpc":"2.0","id":1,"result":{"listenAddresses":["/ip4/127.0.0.1/tcp/48773/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs","/ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs"],"enrUri":"enr:-KO4QCf8bO0zdJNbmM4XrC9-1YmI410uhzxkluhyNrnfP8AvCeHsApydLetCGLNmLAF6_Dcim1Z6CHw3rhWZNfdg22EBgmlkgnY0gmlwhH8AAAGKbXVsdGlhZGRyc4wACgR_AAABBr6G3QOJc2VjcDI1NmsxoQI3RyiMKvo4ZVlO7vyVNuKzwQc4ON6ujhqvGUMVRqVmnIN0Y3CCvoWFd2FrdTIL"}} +8ms
  waku:waku Dialing to /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs with protocols store,filter,lightpush,peer-exchange +154ms
  waku:wait-for-remote-peer /vac/waku/store/2.0.0-beta4 peer found:  16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs +0ms
  waku:store Querying history with the following options { peerId: undefined } +0ms
  waku:libp2p-utils Using codec /vac/waku/store/2.0.0-beta4 +0ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +52ms
  waku:store 5 messages retrieved from store +42ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +60ms
  waku:store 5 messages retrieved from store +58ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +64ms
  waku:store 5 messages retrieved from store +56ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +62ms
  waku:store 5 messages retrieved from store +56ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +58ms
  waku:store 5 messages retrieved from store +54ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +56ms
  waku:store 5 messages retrieved from store +55ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +56ms
  waku:store 5 messages retrieved from store +53ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +57ms
  waku:store 5 messages retrieved from store +54ms
  waku:store Querying store peer /ip4/127.0.0.1/tcp/48774/ws/p2p/16Uiu2HAky9TLTHneWMPEcsgxLH5XzwPxGyYnq7VR6JvGQ5r6siGs for (/waku/2/default-waku/proto) [ '/test/1/waku-store/utf8' ] +55ms
  waku:store 4 messages retrieved from store +55ms
  waku:store Stopping pagination due to `response.pagingInfo.cursor` missing from store response +0ms
[
  14, 15, 16, 17, 18, 12, 13, 14, 15, 16,
  10, 11, 12, 13, 14,  8,  9, 10, 11, 12,
   6,  7,  8,  9, 10,  4,  5,  6,  7,  8,
   2,  3,  4,  5,  6,  0,  1,  2,  3,  4,
   0,  1,  2,  3
]
  waku:nwaku nwaku 1593672 getting SIGINT at 2:11:11 pm +1s
  waku:nwaku nwaku 1593672 interrupted: true +0ms
  waku:nwaku Deleting db file ygyqj.sqlite3 +0ms
  waku:nwaku Failed to clean store file [Error: ENOENT: no such file or directory, stat 'ygyqj.sqlite3'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'stat',
  path: 'ygyqj.sqlite3'
} +25ms
  waku:nwaku nwaku 1593672 process exited with 0 at 2:11:11 pm +2ms

AssertionError: expected 44 to equal 20
Expected :20
Actual   :44
<Click to see difference>

    at Context.<anonymous> (file:///home/fryorcraken/src/waku-org/js-waku/packages/tests/tests/store.node.spec.ts:93:30)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)

```

--- a/tests/v2/waku_archive/test_driver_queue_query.nim
+++ b/tests/v2/waku_archive/test_driver_queue_query.nim
@@ -1,7 +1,7 @@
 {.used.}
 
 import
-  std/[options, sequtils, random],
+  std/[options, sequtils, random, algorithm],
   testutils/unittests,
   chronos,
   chronicles
@@ -160,7 +160,7 @@ suite "Queue driver - query by content topic":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[6..7]
+      filteredMessages == expected[6..7].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -510,7 +510,7 @@ suite "Queue driver - query by cursor":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[2..3]
+      filteredMessages == expected[2..3].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -600,7 +600,7 @@ suite "Queue driver - query by cursor":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[2..5]
+      filteredMessages == expected[2..5].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -706,7 +706,7 @@ suite "Queue driver - query by cursor":
     let expectedMessages = expected.mapIt(it[1])
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expectedMessages[4..5]
+      filteredMessages == expectedMessages[4..5].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -980,7 +980,7 @@ suite "Queue driver - query by time range":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[2..6]
+      filteredMessages == expected[2..6].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -1078,7 +1078,7 @@ suite "Queue driver - query by time range":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[3..4]
+      filteredMessages == expected[3..4].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -1188,7 +1188,7 @@ suite "Queue driver - query by time range":
     let expectedMessages = expected.mapIt(it[1])
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expectedMessages[4..5]
+      filteredMessages == expectedMessages[4..5].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")

--- a/tests/v2/waku_archive/test_driver_sqlite_query.nim
+++ b/tests/v2/waku_archive/test_driver_sqlite_query.nim
@@ -1,7 +1,7 @@
 {.used.}
 
 import
-  std/[options, sequtils, random],
+  std/[options, sequtils, random, algorithm],
   testutils/unittests,
   chronos,
   chronicles
@@ -164,7 +164,7 @@ suite "SQLite driver - query by content topic":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[6..7]
+      filteredMessages == expected[6..7].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -514,7 +514,7 @@ suite "SQLite driver - query by cursor":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[2..3]
+      filteredMessages == expected[2..3].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -604,7 +604,7 @@ suite "SQLite driver - query by cursor":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[2..5]
+      filteredMessages == expected[2..5].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -710,7 +710,7 @@ suite "SQLite driver - query by cursor":
     let expectedMessages = expected.mapIt(it[1])
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expectedMessages[4..5]
+      filteredMessages == expectedMessages[4..5].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -984,7 +984,7 @@ suite "SQLite driver - query by time range":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[2..6]
+      filteredMessages == expected[2..6].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -1082,7 +1082,7 @@ suite "SQLite driver - query by time range":
 
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expected[3..4]
+      filteredMessages == expected[3..4].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")
@@ -1192,7 +1192,7 @@ suite "SQLite driver - query by time range":
     let expectedMessages = expected.mapIt(it[1])
     let filteredMessages = res.tryGet().mapIt(it[1])
     check:
-      filteredMessages == expectedMessages[4..5]
+      filteredMessages == expectedMessages[4..5].reversed()
 
     ## Cleanup
     driver.close().expect("driver to close")

--- a/tests/v2/waku_archive/test_waku_archive.nim
+++ b/tests/v2/waku_archive/test_waku_archive.nim
@@ -12,8 +12,7 @@ import
   ../../../waku/v2/protocol/waku_archive/driver/sqlite_driver,
   ../../../waku/v2/protocol/waku_archive,
   ../../../waku/v2/utils/time,
-  ../testlib/common,
-  ../testlib/switch
+  ../testlib/common
 
 
 proc newTestDatabase(): SqliteDatabase =
@@ -26,6 +25,15 @@ proc newTestArchiveDriver(): ArchiveDriver =
 proc newTestWakuArchive(driver: ArchiveDriver): WakuArchive =
   let validator: MessageValidator = DefaultMessageValidator()
   WakuArchive.new(driver, validator=some(validator))
+
+proc computeTestCursor(pubsubTopic: PubsubTopic, message: WakuMessage): ArchiveCursor =
+  ArchiveCursor(
+    pubsubTopic: pubsubTopic,
+    senderTime: message.timestamp,
+    storeTime: message.timestamp,
+    digest: computeDigest(message)
+  )
+
 
 
 suite "Waku Archive - message handling":
@@ -126,25 +134,25 @@ suite "Waku Archive - message handling":
 procSuite "Waku Archive - find messages":
   ## Fixtures
   let timeOrigin = now()
+  let msgListA = @[
+    fakeWakuMessage(@[byte 00], contentTopic=ContentTopic("2"), ts=ts(00, timeOrigin)),
+    fakeWakuMessage(@[byte 01], contentTopic=ContentTopic("1"), ts=ts(10, timeOrigin)),
+    fakeWakuMessage(@[byte 02], contentTopic=ContentTopic("2"), ts=ts(20, timeOrigin)),
+    fakeWakuMessage(@[byte 03], contentTopic=ContentTopic("1"), ts=ts(30, timeOrigin)),
+    fakeWakuMessage(@[byte 04], contentTopic=ContentTopic("2"), ts=ts(40, timeOrigin)),
+    fakeWakuMessage(@[byte 05], contentTopic=ContentTopic("1"), ts=ts(50, timeOrigin)),
+    fakeWakuMessage(@[byte 06], contentTopic=ContentTopic("2"), ts=ts(60, timeOrigin)),
+    fakeWakuMessage(@[byte 07], contentTopic=ContentTopic("1"), ts=ts(70, timeOrigin)),
+    fakeWakuMessage(@[byte 08], contentTopic=ContentTopic("2"), ts=ts(80, timeOrigin)),
+    fakeWakuMessage(@[byte 09], contentTopic=ContentTopic("1"), ts=ts(90, timeOrigin))
+  ]
+
   let archiveA = block:
       let
         driver = newTestArchiveDriver()
         archive = newTestWakuArchive(driver)
 
-      let msgList = @[
-        fakeWakuMessage(@[byte 00], contentTopic=ContentTopic("2"), ts=ts(00, timeOrigin)),
-        fakeWakuMessage(@[byte 01], contentTopic=ContentTopic("1"), ts=ts(10, timeOrigin)),
-        fakeWakuMessage(@[byte 02], contentTopic=ContentTopic("2"), ts=ts(20, timeOrigin)),
-        fakeWakuMessage(@[byte 03], contentTopic=ContentTopic("1"), ts=ts(30, timeOrigin)),
-        fakeWakuMessage(@[byte 04], contentTopic=ContentTopic("2"), ts=ts(40, timeOrigin)),
-        fakeWakuMessage(@[byte 05], contentTopic=ContentTopic("1"), ts=ts(50, timeOrigin)),
-        fakeWakuMessage(@[byte 06], contentTopic=ContentTopic("2"), ts=ts(60, timeOrigin)),
-        fakeWakuMessage(@[byte 07], contentTopic=ContentTopic("1"), ts=ts(70, timeOrigin)),
-        fakeWakuMessage(@[byte 08], contentTopic=ContentTopic("2"), ts=ts(80, timeOrigin)),
-        fakeWakuMessage(@[byte 09], contentTopic=ContentTopic("1"), ts=ts(90, timeOrigin))
-      ]
-
-      for msg in msgList:
+      for msg in msgListA:
         require driver.put(DefaultPubsubTopic, msg, computeDigest(msg), msg.timestamp).isOk()
 
       archive
@@ -348,120 +356,76 @@ procSuite "Waku Archive - find messages":
       response.messages.anyIt(it == msg3)
 
   test "handle query with forward pagination":
-    ## Setup
-    let
-      driver = newTestArchiveDriver()
-      archive = newTestWakuArchive(driver)
-
-    let currentTime = now()
-    let msgList = @[
-        fakeWakuMessage(@[byte 0], contentTopic=ContentTopic("2"), ts=currentTime - 9),
-        fakeWakuMessage(@[byte 1], contentTopic=DefaultContentTopic, ts=currentTime - 8),
-        fakeWakuMessage(@[byte 2], contentTopic=DefaultContentTopic, ts=currentTime - 7),
-        fakeWakuMessage(@[byte 3], contentTopic=DefaultContentTopic, ts=currentTime - 6),
-        fakeWakuMessage(@[byte 4], contentTopic=DefaultContentTopic, ts=currentTime - 5),
-        fakeWakuMessage(@[byte 5], contentTopic=DefaultContentTopic, ts=currentTime - 4),
-        fakeWakuMessage(@[byte 6], contentTopic=DefaultContentTopic, ts=currentTime - 3),
-        fakeWakuMessage(@[byte 7], contentTopic=DefaultContentTopic, ts=currentTime - 2),
-        fakeWakuMessage(@[byte 8], contentTopic=DefaultContentTopic, ts=currentTime - 1),
-        fakeWakuMessage(@[byte 9], contentTopic=ContentTopic("2"), ts=currentTime)
-      ]
-
-    for msg in msgList:
-      require driver.put(DefaultPubsubTopic, msg, computeDigest(msg), msg.timestamp).isOk()
-
     ## Given
-    var req = ArchiveQuery(
-      contentTopics: @[DefaultContentTopic],
-      pageSize: 2,
+    let req = ArchiveQuery(
+      pageSize: 4,
       ascending: true
     )
 
     ## When
-    var res = archive.findMessages(req)
-    require res.isOk()
+    var nextReq = req # copy
 
-    var
-      response = res.tryGet()
-      totalMessages = response.messages.len()
-      totalQueries = 1
+    var pages = newSeq[seq[WakuMessage]](3)
+    var cursors = newSeq[Option[ArchiveCursor]](3)
 
-    while response.cursor.isSome():
-      require:
-        totalQueries <= 4 # Sanity check here and guarantee that the test will not run forever
-        response.messages.len() == 2
-
-      req.cursor = response.cursor
-
-      # Continue querying
-      res = archive.findMessages(req)
+    for i in 0..<3:
+      let res = archiveA.findMessages(nextReq)
       require res.isOk()
-      response = res.tryGet()
-      totalMessages += response.messages.len()
-      totalQueries += 1
+
+      # Keep query response content
+      let response = res.get()
+      pages[i] = response.messages
+      cursors[i] = response.cursor
+
+      # Set/update the request cursor
+      nextReq.cursor = cursors[i]
 
     ## Then
     check:
-      totalQueries == 4 # 4 queries of pageSize 2
-      totalMessages == 8 # 8 messages in total
+      cursors[0] == some(computeTestCursor(DefaultPubsubTopic, msgListA[3]))
+      cursors[1] == some(computeTestCursor(DefaultPubsubTopic, msgListA[7]))
+      cursors[2] == none(ArchiveCursor)
+
+    check:
+      pages[0] == msgListA[0..3]
+      pages[1] == msgListA[4..7]
+      pages[2] == msgListA[8..9]
 
   test "handle query with backward pagination":
-    ## Setup
-    let
-      driver = newTestArchiveDriver()
-      archive = newTestWakuArchive(driver)
-
-    let currentTime = now()
-    let msgList = @[
-        fakeWakuMessage(@[byte 0], contentTopic=ContentTopic("2"), ts=currentTime - 9),
-        fakeWakuMessage(@[byte 1], contentTopic=DefaultContentTopic, ts=currentTime - 8),
-        fakeWakuMessage(@[byte 2], contentTopic=DefaultContentTopic, ts=currentTime - 7),
-        fakeWakuMessage(@[byte 3], contentTopic=DefaultContentTopic, ts=currentTime - 6),
-        fakeWakuMessage(@[byte 4], contentTopic=DefaultContentTopic, ts=currentTime - 5),
-        fakeWakuMessage(@[byte 5], contentTopic=DefaultContentTopic, ts=currentTime - 4),
-        fakeWakuMessage(@[byte 6], contentTopic=DefaultContentTopic, ts=currentTime - 3),
-        fakeWakuMessage(@[byte 7], contentTopic=DefaultContentTopic, ts=currentTime - 2),
-        fakeWakuMessage(@[byte 8], contentTopic=DefaultContentTopic, ts=currentTime - 1),
-        fakeWakuMessage(@[byte 9], contentTopic=ContentTopic("2"), ts=currentTime)
-      ]
-
-    for msg in msgList:
-      require driver.put(DefaultPubsubTopic, msg, computeDigest(msg), msg.timestamp).isOk()
-
     ## Given
-    var req = ArchiveQuery(
-      contentTopics: @[DefaultContentTopic],
-      pageSize: 2,
-      ascending: false
+    let req = ArchiveQuery(
+      pageSize: 4,
+      ascending: false  # backward
     )
 
     ## When
-    var res = archive.findMessages(req)
-    require res.isOk()
+    var nextReq = req # copy
 
-    var
-      response = res.tryGet()
-      totalMessages = response.messages.len()
-      totalQueries = 1
+    var pages = newSeq[seq[WakuMessage]](3)
+    var cursors = newSeq[Option[ArchiveCursor]](3)
 
-    while response.cursor.isSome():
-      require:
-        totalQueries <= 4 # Sanity check here and guarantee that the test will not run forever
-        response.messages.len() == 2
-
-      req.cursor = response.cursor
-
-      # Continue querying
-      res = archive.findMessages(req)
+    for i in 0..<3:
+      let res = archiveA.findMessages(nextReq)
       require res.isOk()
-      response = res.tryGet()
-      totalMessages += response.messages.len()
-      totalQueries += 1
+
+      # Keep query response content
+      let response = res.get()
+      pages[i] = response.messages
+      cursors[i] = response.cursor
+
+      # Set/update the request cursor
+      nextReq.cursor = cursors[i]
 
     ## Then
     check:
-      totalQueries == 4 # 4 queries of pageSize 2
-      totalMessages == 8 # 8 messages in total
+      cursors[0] == some(computeTestCursor(DefaultPubsubTopic, msgListA[6]))
+      cursors[1] == some(computeTestCursor(DefaultPubsubTopic, msgListA[2]))
+      cursors[2] == none(ArchiveCursor)
+
+    check:
+      pages[0] == msgListA[6..9]
+      pages[1] == msgListA[2..5]
+      pages[2] == msgListA[0..1]
 
   test "handle query with no paging info - auto-pagination":
     ## Setup

--- a/waku/v2/protocol/waku_archive/archive.nim
+++ b/waku/v2/protocol/waku_archive/archive.nim
@@ -5,7 +5,7 @@ else:
 
 
 import
-  std/[tables, times, sequtils, options],
+  std/[tables, times, sequtils, options, algorithm],
   stew/results,
   chronicles,
   chronos,
@@ -129,6 +129,7 @@ proc findMessages*(w: WakuArchive, query: ArchiveQuery): ArchiveResult {.gcsafe.
   if qContentTopics.len > 10:
     return err(ArchiveError.invalidQuery("too many content topics"))
 
+
   let queryStartTime = getTime().toUnixFloat()
 
   let queryRes = w.driver.getMessages(
@@ -145,25 +146,23 @@ proc findMessages*(w: WakuArchive, query: ArchiveQuery): ArchiveResult {.gcsafe.
   waku_archive_query_duration_seconds.observe(queryDuration)
 
 
-
   # Build response
   if queryRes.isErr():
     return err(ArchiveError(kind: ArchiveErrorKind.DRIVER_ERROR, cause: queryRes.error))
 
   let rows = queryRes.get()
 
-  if rows.len <= 0:
-    return ok(ArchiveResponse(
-      messages: @[],
-      cursor: none(ArchiveCursor)
-    ))
-
-
-  # TODO: Move cursor generation to the driver implementation module
-  var messages = if rows.len <= int(qMaxPageSize): rows.mapIt(it[1])
-                 else: rows[0..^2].mapIt(it[1])
+  var messages = newSeq[WakuMessage]()
   var cursor = none(ArchiveCursor)
 
+  if rows.len == 0:
+    return ok(ArchiveResponse(messages: messages, cursor: cursor))
+
+  ## Messages
+  let pageSize = min(rows.len, int(qMaxPageSize))
+  messages = rows[0..<pageSize].mapIt(it[1])
+
+  ## Cursor
   if rows.len > int(qMaxPageSize):
     ## Build last message cursor
     ## The cursor is built from the last message INCLUDED in the response
@@ -171,22 +170,25 @@ proc findMessages*(w: WakuArchive, query: ArchiveQuery): ArchiveResult {.gcsafe.
     let (pubsubTopic, message, digest, storeTimestamp) = rows[^2]
 
     # TODO: Improve coherence of MessageDigest type
-    var messageDigest: array[32, byte]
-    for i in 0..<min(digest.len, 32):
-      messageDigest[i] = digest[i]
+    let messageDigest = block:
+        var data: array[32, byte]
+        for i in 0..<min(digest.len, 32):
+          data[i] = digest[i]
+
+        MessageDigest(data: data)
 
     cursor = some(ArchiveCursor(
       pubsubTopic: pubsubTopic,
       senderTime: message.timestamp,
       storeTime: storeTimestamp,
-      digest: MessageDigest(data: messageDigest)
+      digest: messageDigest
     ))
 
+  # All messages MUST be returned in chronological order
+  if not qAscendingOrder:
+    reverse(messages)
 
-  ok(ArchiveResponse(
-    messages: messages,
-    cursor: cursor
-  ))
+  ok(ArchiveResponse(messages: messages, cursor: cursor))
 
 
 # Retention policy

--- a/waku/v2/protocol/waku_archive/driver/queue_driver/queue_driver.nim
+++ b/waku/v2/protocol/waku_archive/driver/queue_driver/queue_driver.nim
@@ -4,7 +4,7 @@ else:
   {.push raises: [].}
 
 import
-  std/[options, algorithm],
+  std/options,
   stew/results,
   stew/sorted_set,
   chronicles
@@ -279,13 +279,7 @@ method getMessages*(
   if pageRes.isErr():
     return err($pageRes.error)
 
-  var rows = pageRes.value
-
-  # All messages MUST be returned in chronological order
-  if not ascendingOrder:
-    reverse(rows)
-
-  ok(rows)
+  ok(pageRes.value)
 
 
 method getMessagesCount*(driver: QueueDriver): ArchiveDriverResult[int64] =

--- a/waku/v2/protocol/waku_archive/driver/sqlite_driver/sqlite_driver.nim
+++ b/waku/v2/protocol/waku_archive/driver/sqlite_driver/sqlite_driver.nim
@@ -6,7 +6,7 @@ else:
   {.push raises: [].}
 
 import
-  std/[options, algorithm],
+  std/options,
   stew/[byteutils, results],
   chronicles
 import
@@ -107,7 +107,7 @@ method getMessages*(
 ): ArchiveDriverResult[seq[ArchiveRow]] =
   let cursor = cursor.map(toDbCursor)
 
-  var rows = ?s.db.selectMessagesByHistoryQueryWithLimit(
+  let rows = ?s.db.selectMessagesByHistoryQueryWithLimit(
     contentTopic,
     pubsubTopic,
     cursor,
@@ -116,10 +116,6 @@ method getMessages*(
     limit=maxPageSize,
     ascending=ascendingOrder
   )
-
-  # All messages MUST be returned in chronological order
-  if not ascendingOrder:
-    reverse(rows)
 
   ok(rows)
 

